### PR TITLE
[sival,gpio] gpio interrupt test

### PR DESCRIPTION
**Pull request by jdonjdon**
This PR includes

- gpio input mode interrupt test
- gpio output mode interrupt test

Since gpio interrupt only depends on its input, for output mode test, pinmux configured s.t. it loopbacks to the gpio input.
For each mode, four tests are included (Rising edge, Falling edge, level high and level low).

### CODE REVIEW

File: sw/host/tests/chip/gpio_intr/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+fn test_gpio_outputs(opts: &Opts, transport: &TransportWrapper) -> Result<()> {
+    let uart = transport.uart("console")?;
+    log::info!(
+        "Configuring pinmux for {:?}",
+        opts.init.backend_opts.interface
+    );
+    let config = CONFIG
+        .get(opts.init.backend_opts.interface.as_str())
+        .expect("interface");
+    PinmuxConfig::configure(&*uart, None, Some(&config.output))?;
```
Comment (a-will) [inline sw/host/tests/chip/gpio_intr/src/main.rs:242]: If you want to do looped back / self-triggered interrupt stimulus, then the inputs will need to be connected too:

```suggestion
    PinmuxConfig::configure(&*uart, Some(&config.input), Some(&config.output))?;
```

File: sw/device/tests/gpio_intr_test.c [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTTF_DEFINE_TEST_CONFIG(.enable_uart_flow_control = true);
+
+static dif_gpio_t gpio;
+static dif_pinmux_t pinmux;
+static dif_uart_t uart0;
+static dif_rv_plic_t plic;
+static uint32_t expected_intr[] = {17, 18, 19, 20, 21, 22, 25, 26, 27, 28};
```
Comment (engdoreis) [inline sw/device/tests/gpio_intr_test.c:34]:
```suggestion
static const uint32_t kExpectedIntr[] = {17, 18, 19, 20, 21, 22, 25, 26, 27, 28};
```

File: sw/host/tests/chip/gpio_intr/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+        );
+    } else {
+        let write_val;
+        // invert == 1: use for walking 0
+        // invert == 0: use for walking 1
+        if invert == 1 {
+          write_val = !masked_value;
+        } else {
+          write_val = masked_value;
+        }
```
Comment (engdoreis) [inline sw/host/tests/chip/gpio_intr/src/main.rs:121]: nit.
```suggestion
        let write_val = if invert == 1 { !masked_value} else { masked_value};
```

File: sw/host/tests/chip/gpio_intr/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    for pin in config.output.keys() {
+        transport
+            .gpio_pin(&pin.to_string())?
+            .set_mode(PinMode::Input)?;
+    }
+
+    log::info!("Enabling outputs on the DUT");
+
+    // Output Rising Edge test
+    GpioSet::irq_set_trigger(&*uart, 0xFFFFFFFF, "rising".into())?;
```
Comment (engdoreis) [inline sw/host/tests/chip/gpio_intr/src/main.rs:204]:
```suggestion
    GpioSet::irq_set_trigger(&*uart, u32::MAX, "rising".into())?;
```

**Review by engdoreis — APPROVED** [review]

**Review by engdoreis — APPROVED** [review]
LGTM. Can you also link this test on the testplan?

File: sw/device/tests/gpio_intr_test.c (hunk 2 above)
Comment (jdonjdon) [reply sw/device/tests/gpio_intr_test.c]: updated

File: sw/host/tests/chip/gpio_intr/src/main.rs (hunk 4 above)
Comment (jdonjdon) [reply sw/host/tests/chip/gpio_intr/src/main.rs]: updated

**Comment by jdonjdon** [comment]
> LGTM. Can you also link this test on the testplan?

sure, testplan is updated.

**Pull request merged**